### PR TITLE
feat: refactor table component

### PR DESCRIPTION
**Pull request by chrishiguto**
### Context and reasoning behind the refactor

While Table is a great component with good defaults, extending and customizing is hard. Almost every time that we had to work with the Table component in different projects we had to perform changes to rockets-react.
Currently, the Table component relies on an inheritance pattern that uses React props to render different variants. Although this approach has its advantages, it complicates customization and the developer experience.
rockets-react components should be minimally opinionated and highly extendable. This is because we can't anticipate the requirements of future projects using rockets-react.

The idea is to avoid locking in the developer and allowing customization of specific component parts. Exposing smaller, minimally functional components will provide out-of-the-box functionality while still allowing for customized rendering output to meet project requirements.

We need a Table component with sensible defaults and easy extensibility, without the need to toggle multiple props, ensuring code remains both readable and concise.

Our goal is not to refactor every single line of code. Instead, we aim to simplify much of the existing code. You may still encounter old code that could be refactored in the future.

## New features

- `Table.Root` is the highest components of the tree and holds the context to share state. It can function as either a controlled or uncontrolled component. When passing a `tableQueryState` to `Table.Root`, it indicates the intention to control the table's state. This is particularly helpful when managing data externally and requiring full control. `Table.Root `intelligently enforces the requirement of `total`, `pageCount`, and `updateTableQueryState` if any of these props are provided. In the absence of these props, the component remains uncontrolled, and the state is managed by the `Table` component.

```
<Table.Root rows={customRows()} headers={headers} > // This works and the state is controlled by the component

 // This also works and it means we're controlling the state outside of the component
// If any of the additional props are passed, all others are required.
<Table.Root 
   rows={customRows()} 
   headers={headers} 
   tableQueryState={...} // I'm passing this, so every prop below are required
   updateTableQueryState={...}
   total={...}
   pageCount={...}
>
```

- Instead of exposing a single Table component, we now expose each part of the `Table` component individually and integrate them using a custom hook called useTableRoot to share state.

- While we provide out-of-the-box components, we utilize render props in `Table.BodyRows ` to manage pagination, sorting, filtering, and more. However, we still grant developers the flexibility to use their own components for row rendering.

```
...
<Table.BodyRows
   renderRow={row => (
      <Table.BodyRow row={row} hasCheckboxes={false} hover={false}>
         <Table.BodyCell row={row} />
      </Table.BodyRow>
   )} 
/>
...
```

- You can create custom components and use the `useTableRoot` hook to access the necessary data from the `Table` state.

```
...
const MyCustomComponent = () => {
   const state = useTableRoot();

   return (...)
}
...
...
```

- We now employ URL state for both manual pagination (server-side) and local pagination (client-side).
- We offer a function called `createTableTheme` to facilitate customizing the table layout without the need to create extensive objects directly in JSX/TSX. This enhances code readability and simplifies style toggling.

```
...
  const tableTheme = createTableTheme({
    table: {
      height: '100%',
    },
    root: {
      display: 'flex',
      flexDirection: 'column',
      flex: 1,
      overflow: 'auto',
    },

 <Table.Root sx={tableTheme.root}>
    <Table.Table stickyHeader variant="outlined" sx={tableTheme.table}>
   ...
```

- `useTable` now uses the `useTableQueryState` hook, creating a state object to manage pagination, filtering, and sorting. Importantly, `useTable` is now decoupled from the `Table` component and can function independently. It implements the same interface for the query state. based on the It can receive an initial state, or if one is not provided, it will utilize the default `useTableQueryState` settings. It exposes the fetched data, the current state, and provides methods for updating the state.

- Every component now includes JSDocs, and even some inner functions have JSDocs as well.

## Full example

```
const tableTheme = createTableTheme({
    table: {
      height: '100%',
    },
    root: {
      display: 'flex',
      flexDirection: 'column',
      flex: 1,
      overflow: 'auto',
    },
    tableHeader: {
      ...theme.typography.caption,
      lineHeight: 1,
      fontWeight: 500,
      color: theme.palette.grey[500],
    },
    tableRow: {
      backgroundColor: '#F9FAFB',
      textTransform: 'uppercase',
    },
    tableContainer: {
      flex: 1,
    },
  });

<Table.Root rows={customRows()} headers={headers} sx={tableTheme.root} >
   <TableContainer sx={tableTheme.tableContainer}>
      <Table.Table stickyHeader variant="outlined" sx={tableTheme.table}>
         <TableHead>
            <TableRow sx={tableTheme.tableRow}>
               <Table.HeaderCell sx={tableTheme.tableHeader} />
            </TableRow>
         </TableHead>
         <TableBody>
            <Table.BodyRows
               renderRow={row => (
                  <Table.BodyRow row={row} hasCheckboxes={false} hover={false}>
                     <Table.BodyCell row={row} />
                  </Table.BodyRow>
               )}
             />
         </TableBody>
      </Table.Table>
   </TableContainer>
   <Table.Pagination variant="outlined" />
</Table.Root>
```

### CODE REVIEW

File: packages/react-material-ui/src/components/Table/TableStyles/MSWTableStyles.ts
Comment (chrishiguto) [inline packages/react-material-ui/src/components/Table/TableStyles/MSWTableStyles.ts:1]: I think this file should be local to MSW project

File: packages/react-material-ui/src/components/Table/TableRoot.tsx [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  ) => {
+    const selectedIndex = selected.findIndex((_row) => _row.id === row.id);
+    let newSelected: RowProps[] = [];
+
+    if (selectedIndex === -1) {
+      newSelected = newSelected.concat(selected, row);
+    } else if (selectedIndex === 0) {
+      newSelected = newSelected.concat(selected.slice(1));
+    } else if (selectedIndex === selected.length - 1) {
+      newSelected = newSelected.concat(selected.slice(0, -1));
```
Comment (andreneto97) [inline packages/react-material-ui/src/components/Table/TableRoot.tsx:133]: This two is for performance?

Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableRoot.tsx:133]: > Our goal is not to refactor every single line of code. Instead, we aim to simplify much of the existing code. You may still encounter old code that could be refactored in the future.


This code already exists in the codebase today and this also comes from the official docs, please check: https://mui.com/material-ui/react-table/#sorting-amp-selecting

File: packages/react-material-ui/src/components/Table/TableHeader/TableHeaderCell.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  const { order, orderBy } = tableQuery;
+
+  /**
+   * Creates a sort handler function for a specific table column.
+   *
+   * @param property - The property to sort by.
+   * @param event - The mouse event triggering the sorting.
+   * @returns {void} - This function doesn't return anything.
+   */
+  const createSortHandler =
```
[the remark this answers is not in the document]
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableHeader/TableHeaderCell.tsx:34]: > Our goal is not to refactor every single line of code. Instead, we aim to simplify much of the existing code. You may still encounter old code that could be refactored in the future.

This code already exists in the codebase today and this also comes from the official docs, please check: https://mui.com/material-ui/react-table/#sorting-amp-selecting

File: packages/react-material-ui/src/components/Table/TableBody/TableBodyCells.tsx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ */
+export const TableBodyCells = ({ row, ...rest }: TableBodyCellsProps) => {
+  const { headers } = useTableRoot();
+
+  return (
+    <>
+      {headers.map((hd) => {
+        return (
+          <TableCell key={hd.id} {...rest}>
+            {getCellData(row[hd.id])}
```
[the remark this answers is not in the document]
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableBody/TableBodyCells.tsx:59]: This code already exists in the codebase today and it is using `id`, not `key`. We don't even have a `key` property in `HeadersProps` type.

Where is `key` coming from? Are you sure about this?

File: packages/react-material-ui/src/components/Table/TableBody/TableBodyCells.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        {cell.value ?? ''}
+      </Text>
+    );
+  } else {
+    return (
+      <Tooltip title={cell.title}>
+        <span>{cell.value ?? ''}</span>
+      </Tooltip>
+    );
+  }
```
[the remark this answers is not in the document]
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableBody/TableBodyCells.tsx:38]: `Our goal is not to refactor every single line of code. Instead, we aim to simplify much of the existing code. You may still encounter old code that could be refactored in the future.`

This code already exists in the codebase today.

File: packages/react-material-ui/src/components/Table/TableRoot.tsx [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+  ...rest
+}: PropsWithChildren<TableRootProps & BoxProps>) => {
+  const searchParams = useSearchParams();
+  const pathname = usePathname();
+  const router = useRouter();
+  const { tableQueryState, setTableQueryState } = useTableQueryState();
+
+  const [selected, setSelected] = useState<RowProps[]>([]);
+
+  const isControlled = !!controlledTableQueryState;
```
[the remark this answers is not in the document]
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableRoot.tsx:54]: Can you explain this a little further?

`controlledTableQueryState` is an object and if the user passes it we're setting the value of `isControlled` to true. This is used if different parts of the TableRoot to know wether the data is managed internally our externally.

```
<Table.Root rows={customRows()} headers={headers} > // This works and the state is controlled by the component

 // This also works and it means we're controlling the state outside of the component
// If any of the additional props are passed, all others are required.
<Table.Root 
   rows={customRows()} 
   headers={headers} 
   tableQueryState={...} // I'm passing this, so every prop below are required
   updateTableQueryState={...}
   total={...}
   pageCount={...}
>
```

File: packages/react-material-ui/src/components/Table/TableHeader/TableHeaderCell.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        >
+          {cell.label}
+          {orderBy === cell.id ? (
+            <Box component="span" sx={visuallyHidden}>
+              {order === 'desc' ? 'sorted descending' : 'sorted ascending'}
+            </Box>
+          ) : null}
+        </TableSortLabel>
+      ) : (
+        <>{cell.label}</>
```
[the remark this answers is not in the document]
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableHeader/TableHeaderCell.tsx:63]: This code already exists in the codebase today. This is just so the user can decide wether to have sorting or not, disabling the table sort label.

I can remove the Fragment though.

File: packages/react-material-ui/src/components/Table/TableBody/TableBodyCells.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * @param {TableBodyCellsProps} props - The props for the TableBodyCells component.
+ * @returns A React element representing the table body cell.
+ */
+export const TableBodyCells = ({ row, ...rest }: TableBodyCellsProps) => {
+  const { headers } = useTableRoot();
+
+  return (
+    <>
+      {headers.map((hd) => {
```
[the remark this answers is not in the document]
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableBody/TableBodyCells.tsx:56]: Done

File: packages/react-material-ui/src/components/Table/TableBody/TableBodyRows.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    (row, index) => {
+      const labelId = `table-checkbox-${index}`;
+
+      if (!renderRow) {
+        return renderDefaultRow(row);
+      }
+
+      return renderRow(row, labelId);
+    },
+  );
```
[the remark this answers is not in the document]
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/TableBody/TableBodyRows.tsx:79]: Done

File: packages/react-material-ui/src/components/Table/hooks/useTableRoot.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  headers: HeaderProps[];
+  total: number;
+  isControlled: boolean;
+  pageCount: number;
+  tableQuery: TableQueryStateProps;
+  selected: RowProps[];
+  isSelected: (id: string) => boolean;
+  handleChangePage: (event: unknown, newPage: number) => void;
+  handleChangeRowsPerPage: (event: React.ChangeEvent<HTMLInputElement>) => void;
+  handleRequestSort: (
```
Comment (brunomous) [inline packages/react-material-ui/src/components/Table/hooks/useTableRoot.tsx:17]: Maybe the name of this attribute should be more explicit. What does `handleRequestSort` mean? If it's sorting by a request response, wouldn't `handleSortByRequestData` be a better name?

File: packages/react-material-ui/src/components/Table/TableStyles/MSWTableStyles.ts
Comment (andreneto97) [reply packages/react-material-ui/src/components/Table/TableStyles/MSWTableStyles.ts:1]: We agreed to remove

File: packages/react-material-ui/src/components/Table/hooks/useTableRoot.tsx (hunk 9 above)
Comment (chrishiguto) [reply packages/react-material-ui/src/components/Table/hooks/useTableRoot.tsx:17]: This comes from the official material docs. Though in this case it's not related to a network request so I agree with you.

rename to `handleSort`.

**Pull request merged**